LFS-996: Fine tune vocabulary search

veronikaslc opened the pull request:
Subtasks:
- [x] [LFS-1105: Display vocabularies in a MaterialTable](https://phenotips.atlassian.net/browse/LFS-1105)
- [x] [LFS-1106: Turn the current vocabulary search bar into Advanced search](https://phenotips.atlassian.net/browse/LFS-1106)
- [x] [LFS-1113: Hide the "Bioportal API Key" form when the key is set and queries are working, to avoid distracting the user](https://phenotips.atlassian.net/browse/LFS-1113)
- [x] [LFS-1121: Adjust vertical spacing on the Vocabularies page](https://phenotips.atlassian.net/browse/LFS-1121)

Commit:
LFS-1105: Display vocabularies in a MaterialTable

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -87,55 +63,81 @@ export default function VocabularyTable(props) {
+                cellStyle: {
+                  width: '10%',
+                  whiteSpace: "pre",
+                  fontWeight: '400'
```
Specifying `fontWeight` seems unnecessary for all columns
```suggestion
```

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -87,55 +63,81 @@ export default function VocabularyTable(props) {
+              { title: 'Name',
+                cellStyle: {
+                  width: "33%",
+                  whiteSpace: "pre",
+                  fontWeight: '400'
```
I think it's ok if the vocab name is not wrapped
```suggestion
```

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -87,55 +63,81 @@ export default function VocabularyTable(props) {
+                cellStyle: {
+                  width: "20%",
+                  whiteSpace: "pre",
+                  fontWeight: '400'
```
```suggestion
```

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -87,55 +63,81 @@ export default function VocabularyTable(props) {
+                cellStyle: {
+                  width: "14%",
+                  whiteSpace: "nowrap",
+                  fontWeight: '400'
```
```suggestion
```

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -17,56 +17,32 @@
+  const { vocabList, type  } = props;
+  const [filterTable, setFilterTable] = useState(false);
+  const [acronymList, setAcronymList] = useState([]);
+  const [ rowCount, setRowCount ] = useState(5);
```
I think we should show more per page
```suggestion
  const [ rowCount, setRowCount ] = useState(10);
```

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -87,55 +63,81 @@ export default function VocabularyTable(props) {
+                           },
+            }}
+            localization={{
+              toolbar : { searchTooltip: "Filter",
```
I feel like this should be more explicit. How about:
```suggestion
              toolbar : { searchTooltip: "Filter by vocabulary id or name ",
```

marta- reviewed (commented):
This is great. Just some minor tweaks needed.

marta- committed:
LFS-1105: Display vocabularies in a MaterialTable

Apply suggestions from code review

marta- commented on modules/vocabularies/src/main/frontend/src/vocabularyTable.jsx:
```diff
@@ -87,55 +63,76 @@ export default function VocabularyTable(props) {
+            data={vocabList}
+            options={{
+              search: true,
+              exportButton: true,
```
We don't need an export button
```suggestion
```

marta- committed:
LFS-1105: Display vocabularies in a MaterialTable
* Removed export button
* Left-aligned table column titles
* Removed unnecessary (and looking slightly misplaced if left-aligned) "Actions" column title

marta- committed:
LFS-1105: Display vocabularies in a MaterialTable
Fixed regression: long "version" field causes extra-wide table.
Examples of ontologies with this issue: LICO, FISHO
Also: displayed version in the "About" dialog

marta- committed:
LFS-1105: Display vocabularies in a MaterialTable
* Removed unnecessary "Browse..." heading that was displayed between the "Search by keywords" bar and the table
* Removed MaterialTable's search bar to avoid visual redundancy with the "Search by keywords" bar
* Replaced the search functionality with filtering on the Id and Name columns
* Also enabled sorting by Id and Name

Commit:
LFS-1105: Display vocabularies in a MaterialTable
Removed default table option

sashaandjic commented:
- search does not work; cannot locate HANCESTRO, HP,...; after the search is complete INSTALL and ABOUT buttons are no longer visible
- enlarge the size of the font in the table header to match the size in the detail
![image](https://user-images.githubusercontent.com/23243244/115264386-4258f780-a104-11eb-8e5f-d2993a8cd55a.png)

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (dismissed):
This is buggy.

- filter by "ancestr" in the ID columnt to get HANCESTRO displayed, press Install; after it gets installed the filter is discarded, the first page is displayed, and OGG shows up as INSTALLED
- then I filtered by "HP", got it installed -> after it installed, the filters reset, and this time OGG did not show as installed, however the Installed table only showed HP as installed (I can't reliably get this to happen though, most times bot HP and HANCESTRO correctly show up in the Installed table)
- filter by "HP" again, uninstall -> filters get reset again
- search (by keyword, not filter) "microcephaly" -> most rows lack the action buttons, but a random few might still have them -> no way to install the found vocabularies
- after installing and uninstalling more vocabularies, I managed to get ECO (which I did install) show up in the Installed table, but with an "Install" button, while also showing in the Bioportal table with an "Uninstall" button
- Don't know how, but got these JS errors logged:

```Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
    in VocabularyActions (created by MTableCell)
    in td (created by ForwardRef(TableCell))
    in ForwardRef(TableCell) (created by WithStyles(ForwardRef(TableCell)))
    in WithStyles(ForwardRef(TableCell)) (created by MTableCell)
    in MTableCell (created by MTableBodyRow)
    in tr (created by ForwardRef(TableRow))
    in ForwardRef(TableRow) (created by WithStyles(ForwardRef(TableRow)))
    in WithStyles(ForwardRef(TableRow)) (created by MTableBodyRow)
    in MTableBodyRow (created by MTableBody)
...
    setPhase webpack://lfs-aggregated-frontend/./src/vocabulariesAdminPage.jsx?:153
    setPhase webpack://lfs-aggregated-frontend/./src/vocabularyTable.jsx?:159
    install webpack://lfs-aggregated-frontend/./src/vocabularyActions.jsx?:121
```

and 

```
Warning: Failed prop type: Material-UI: The page prop of a TablePagination is out of range (0 to 0, but page is 1).
    in ForwardRef(TablePagination) (created by WithStyles(ForwardRef(TablePagination)))
    in WithStyles(ForwardRef(TablePagination)) (created by MaterialTable)
    in tr (created by ForwardRef(TableRow))
    in ForwardRef(TableRow) (created by WithStyles(ForwardRef(TableRow)))
    in WithStyles(ForwardRef(TableRow)) (created by MaterialTable)
    in tfoot (created by ForwardRef(TableFooter))
    in ForwardRef(TableFooter) (created by WithStyles(ForwardRef(TableFooter)))
    in WithStyles(ForwardRef(TableFooter)) (created by MaterialTable)
    in table (created by ForwardRef(Table))
    in ForwardRef(Table) (created by WithStyles(ForwardRef(Table)))
    in WithStyles(ForwardRef(Table)) (created by MaterialTable)
    in div (created by ForwardRef(Paper))
    in ForwardRef(Paper) (created by WithStyles(ForwardRef(Paper)))
    in WithStyles(ForwardRef(Paper)) (created by Container)
    in Container (created by MaterialTable)
    in Provider (created by App)
    in App (created by ErrorBoundary)
    in ErrorBoundary (created by DragDropContext)
    in DragDropContext (created by MaterialTable)
    in MaterialTable
    in Unknown (created by WithStyles(Component))
    in WithStyles(Component) (created by VocabularyTable)
    in div (created by ForwardRef(Grid))
    in ForwardRef(Grid) (created by WithStyles(ForwardRef(Grid)))
    in WithStyles(ForwardRef(Grid)) (created by VocabularyTable)
    in VocabularyTable (created by VocabularyDirectory)
    in VocabularyDirectory (created by VocabulariesAdminPage)
    in div (created by ForwardRef(Grid))
    in ForwardRef(Grid) (created by WithStyles(ForwardRef(Grid)))
    in WithStyles(ForwardRef(Grid)) (created by VocabulariesAdminPage)
    in VocabulariesAdminPage (created by Context.Consumer)
    in Page (created by Context.Consumer)
    in Route (created by Main)
    in Switch (created by Main)
    in div (created by Main)
    in div (created by Main)
    in div (created by Main)
    in Suspense (created by Main)
    in div (created by Main)
    in ThemeProvider (created by Main)
    in Main (created by WithStyles(Main))
    in WithStyles(Main) (created by Context.Consumer)
    in Route
    in Switch
    in Router
...
        setState
    MaterialTable webpack://lfs-aggregated-frontend/./node_modules/material-table/dist/material-table.js?:428
    later webpack://lfs-aggregated-frontend/./node_modules/debounce/index.js?:27
```

IMHO, must fix:
- buttons missing when searching

Should fix:

- OGG showing up as installed after installing HANCETRO

Nice to have fixed, but not crucial:

- losing filters after installing a vocabulary
- JS errors
- figure out why after installing both HANCETRO and HP only HP showed up in the Installed table, if it's easy to reproduce
- figure out why ECO shows up as installed but not installed in the "Installed" table

sdumitriu commented:
I have a pretty reliable way of getting an installed vocabulary to show up as uninstalled:
- install ECO, HANCESTRO
- reload the page
- uninstall ECO -> HANCESTRO shows up with INSTALL button
- if you press on the Install button there, you also get the first JS error I mentioned

Commit:
LFS-1105: Display vocabularies in a MaterialTable
Force phase set to initial on props change for VocabularyActions component

Commit:
LFS-1105: Display vocabularies in a MaterialTable
Fixed regression for advanced search table filtering

veronikaslc commented:
Fixed:
- [ ]  buttons missing when searching
- [ ]  installed vocabulary to show up as uninstalled
Vote to leave as expected behaviour after completed action:
 - losing filters after installing a vocabulary
Could not reproduce:
 - all other issues

marta- commented:
> Vote to leave as expected behaviour after completed action:
> 
> * losing filters after installing a vocabulary

I'm okay with merging with lost filters since it's not a blocker, but it should not stay as the default behavior.

Commit:
LFS-1106: Turn the current vocabulary search bar into Advanced search

Commit:
LFS-1106: Turn the current vocabulary search bar into Advanced search
Better looks for the search input

Commit:
LFS-1113: Hide the Bioportal API Key form when the key is set and queries are working, to avoid distracting the user

marta- commented on modules/vocabularies/src/main/frontend/src/bioportalApiKey.jsx:
```diff
@@ -83,58 +97,62 @@ export function BioPortalApiKey(props) {
+      </Typography>
+
+      { !bioPortalApiKey && <>
+         <Grid item className={classes.noKeyInfo}>
```
This creates a `Grid item` inside `Grid item`.

marta- replied:
Also there doesn't seem to be a way to actually save a key when no key is set.

![image](https://user-images.githubusercontent.com/651980/115563548-ab1bad80-a285-11eb-9422-cea12a98e09f.png)

Commit:
LFS-1113: Hide the Bioportal API Key form when the key is set and queries are working, to avoid distracting the user
Added save button, fixed some alignments

marta- commented on modules/vocabularies/src/main/frontend/src/search.jsx:
```diff
@@ -154,45 +112,32 @@ export default function Search(props) {
+          }}
+          label="Search BioPortal by keywords"
+          helperText={ <>
+                         {loading && <LinearProgress/>}
```
This line causes a console warning: `Warning: validateDOMNesting(...): <div> cannot appear as a descendant of <p>.`

Also, the progress bar appears slightly lower than the input bottom and is too narrow, making the whole thing look unpolished.

![image](https://user-images.githubusercontent.com/651980/115595381-d95cb580-a2a4-11eb-9252-1ed5d3c5dd02.png)

EDIT: The linear progress will be replaced with a circular progress around the search button

marta- commented on modules/vocabularies/src/main/frontend/src/bioportalApiKey.jsx:
```diff
@@ -83,58 +97,77 @@ export function BioPortalApiKey(props) {
+        <Typography variant="h6">
+          Find on <a href="https://bioportal.bioontology.org/" target="_blank">BioPortal</a>
+          { bioPortalApiKey &&
+            <Tooltip title="Change BioPortal API key">
+              <IconButton onClick={() => {setDisplayPopup(true)}} >
+                <SettingsIcon/>
+              </IconButton>
+            </Tooltip>
```
The button looks slightly misaligned. I managed to make it look better by adding `margin-top: -4px`.

marta- commented on modules/vocabularies/src/main/frontend/src/bioportalApiKey.jsx:
```diff
@@ -83,58 +97,77 @@ export function BioPortalApiKey(props) {
+        </Typography>
+      </Grid>
+
+      { !bioPortalApiKey && <>
+         <Grid item className={classes.noKeyInfo}>
```
This does not make the remote vocab table disappear and the whole page looks pretty bad with both API key info and vocabularies table at the same time. Also the search doesn't actually work if the key is missing, but there's no feedback to the user it just keeps `loading`.

![image](https://user-images.githubusercontent.com/651980/115596069-ad8dff80-a2a5-11eb-88e4-6c2e10d97caf.png)

Two options:

1. implement the [current specs](https://phenotips.atlassian.net/browse/LFS-1113) and remove the table when the key is missing:
> If there's no environment variable key, deleting the key in this form should result in:
> - the disappearance of the bioportal (remote) vocab table and
> - the display of the empty API key form directly in the Vocabularies page under “Find on BioPortal” (as specified in the first item of this todo list)
2. change to keep hiding the key form and only display the settings icon button if there's already data to display + ensure search gives the appropriate error message.

I still prefer option 1 so the user clearly understands the consequences of deleting the key.

Commit:
LFS-1113: Hide the Bioportal API Key form when the key is set and queries are working, to avoid distracting the user
Better setting button alignment

Commit:
LFS-1113: Hide the Bioportal API Key form when the key is set and queries are working, to avoid distracting the user
Hide remote table if no bioportal key

Commit:
LFS-1106: Turn the current vocabulary search bar into Advanced search
Switch linear bar to circular progress around the search button; Show table reloading status

Commit:
LFS-1106: Turn the current vocabulary search bar into Advanced search
Show empty table if advanced search returned 0 results

marta- committed:
LFS-1121: Adjust vertical spacing on the Vocabularies page

marta- committed:
LFS-1105: Display vocabularies in a MaterialTable
* Reverted to to 5 items per page, otherwise the "loading" state of the table is usually not visible on the screen
* Relabeled "Id" as "Identifier"

marta- committed:
LFS-1106: Turn the current vocabulary search bar into Advanced search
Fixed typo in the helper text

Commit:
LFS-1106: Turn the current vocabulary search bar into Advanced search
Reset filters upon new andvanced search request

marta- committed:
LFS-1106: Turn the current vocabulary search bar into Advanced search
Minor UI and naming adjustments

sashaandjic commented:
Look really nice. Search seems to return no more than 25 entries. Is this by design? 
![image](https://user-images.githubusercontent.com/23243244/115853994-c73f5c00-a3f7-11eb-8ab1-83becc77a952.png)

marta- commented:
> Look really nice. Search seems to return no more than 25 entries. Is this by design?

Good catch. It may be a default limit set in the bioportal api. @veronikaslc please investigate (http://data.bioontology.org/documentation).

However, It's not a regression, it was exactly like this before without us noticing it.

sashaandjic reviewed (approved):
tested
